lib/backend/memory/memory.go

```diff
@@ -374,7 +382,69 @@ func (m *Memory) CompareAndSwap(ctx context.Context, expected backend.Item, repl
 	if !m.EventsOff {
 		m.buf.Emit(event)
 	}
-	return m.newLease(replaceWith), nil
+	return backend.NewLease(replaceWith), nil
+}
+
+func (m *Memory) ConditionalDelete(ctx context.Context, key []byte, rev string) error {
+	if len(key) == 0 {
+		return trace.BadParameter("missing parameter key")
+	}
+	m.Lock()
+	defer m.Unlock()
+	m.removeExpired()
+
+	item, found := m.tree.Get(&btreeItem{Item: backend.Item{Key: key}})
+	if !found {
+		return trace.NotFound("key %q is not found", string(key))
+	}
+
+	if item.Item.Revision != rev {
+		return trace.Wrap(backend.ErrIncorrectRevision)
+	}
+
+	event := backend.Event{
+		Type: types.OpDelete,
+		Item: backend.Item{
+			Key: key,
+		},
+	}
+	m.processEvent(event)
+	if !m.EventsOff {
+		m.buf.Emit(event)
+	}
+	return nil
+}
+
+func (m *Memory) ConditionalUpdate(ctx context.Context, i backend.Item) (*backend.Lease, error) {
+	if len(i.Key) == 0 {
+		return nil, trace.BadParameter("missing parameter key")
+	}
+	m.Lock()
+	defer m.Unlock()
+	m.removeExpired()
+
+	item, found := m.tree.Get(&btreeItem{Item: i})
+	if !found {
+		return nil, trace.NotFound("key %q is not found", string(i.Key))
```

Comment:
Returning a mix of `CompareFailed` and `NotFound` complicates error handling IMO.  The default error message specifically mentions concurrent deletion as a possible source of the error, so I'm not sure there's much benefit from a UX perspective.

More importantly, I think there should be a simple check for "failure due to concurrent modification" that works universally.  Otherwise it will be too easy to accidentally introduce subtle bugs into logic that handles updates.